File: content/chapters/io/lab/support/send-receive/receiver.py
# SPDX-License-Identifier: BSD-3-Clause

import socket  # Exposes the Berkely Sockets API
from os import getpid

IP = "127.0.0.1"  # Loopback IP address (localhost)
MAX_MESSAGE_LENGTH = 1024
PORT = 5000  # Port to listen for data data on (non-privileged ports are > 1023)


def main():
    # Create a socket. Do not mind the arguments for now.
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

    # Bind the socket to the IP and port. This registers the current process as
    # "listening" for data on the port.
    sock.bind((IP, PORT))

    print(f"Server with PID {getpid()} is listening for data on port {PORT}...")

    while True:
        # Receive at most 1024 bytes data from the socket.
        message, (ip, port) = sock.recvfrom(MAX_MESSAGE_LENGTH)

        # The sender had to encode the message from string to bytes. Now decode
        # it back to a string.
        message = message.decode()

        # Print the message.
        print(f"Received from IP {ip} and port {port}: {message}")

        # If the message is "exit", break out of the loop.
        if message == "exit":
            break

    sock.close()


if __name__ == "__main__":
    main()

File: content/chapters/io/lab/support/send-receive/receiver.c
// SPDX-License-Identifier: BSD-3-Clause

#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "sock-utils.h"

const char *localhost = "127.0.0.1";

int main(void)
{
	char buf[BUFSIZE];
	int ret, sockfd;
	struct sockaddr_in receiver_addr;
	struct sockaddr_in sender_addr;
	socklen_t sender_addr_len;

	/* Create a socket. */
	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
	DIE(sockfd < 0, "socket");

	/* TODO 1: Add address and port to receiver_addr. */
	/* HINT: use populate_sockaddr(). */

	/* TODO 2: Bind sockfd to receiver_addr. Use DIE to check for errors. */

	printf("Server with PID %d is listening for data on port %d ...\n", getpid(), PORT);

	while (1) {
		/* TODO 2.1: Remove break. */
		break;

		/* Prepare the buffer */
		memset(buf, 0, BUFSIZE);

		/* TODO 3: Receive at most BUFSIZE bytes through the socket with recvfrom(). */
		/* HINT: Use sender_addr as the fifth arg. You will need to cast it. */
		/*       It will be populated with information about the sender. */
		/* Use DIE to check for errors. */

		/* Ensure buf is null-terminated. */
		buf[BUFSIZE-1] = '\0';

		/* Print the message. */
		printf("Received from IP %s and port %d: %s\n", inet_ntoa(sender_addr.sin_addr),
			sender_addr.sin_port, buf);

		if (strncmp(buf, "exit", 4) == 0)
			break;
	}

	/* Close the socket. */
	ret = close(sockfd);
	DIE(ret < 0, "close");

	return 0;
}
